src/UiModes.ts

```diff
@@ -0,0 +1,321 @@
+            } )
+
+        })
+    }
+
+    static updateSettingsArray = () : void => {
+        UiModeSystem.getActiveUiMode().getSettings().forEach(function(setting){
+            Setting.setValue(setting.getKey(),setting.getValue())
+        })
+    }
+
+    static getActiveSetting = (settingName:string) : any => {
+        let found = false
+        let value:any = null
+        UiModeSystem.getActiveUiMode().getSettings().forEach(function(setting){
+            if(setting.getKey() === settingName){
+                value = setting.getValue();
+                found = true;
+            }
+        })
+        if(!found){
+            console.warn('Requested setting: "'+ settingName+'" can not be found')
+        }else{
+            return value
+        }
+    }
+
+    static setActiveSetting = (settingName:string,newValue:any) : any => {
+        let found = false
+        let settingIsPerpetual = false
+
+        //if a setting is marked perpetual we will write the value to all ui modes, this means it stayes the same regardless of which ui  mode is active
+        UiModeSystem.getActiveUiMode().getSettings().forEach(function(setting){
+            if(setting.getKey() === settingName){
+                settingIsPerpetual = setting.isPerpetural()
+            }
+        })
+
+        if(settingIsPerpetual){
+                UiModeSystem.getUiModes().forEach(function(uiMode){
+                    uiMode.getSettings().forEach(function(setting){
+                        if(setting.getKey() === settingName){
+                            found = true
+                            setting.setValue(newValue)
+                        }
+                    })
+                })
+        }else{
+            UiModeSystem.getActiveUiMode().getSettings().forEach(function(setting){
+                if(setting.getKey() === settingName){
+                    found = true
+                    setting.setValue(newValue)
+                }
+            })
+        }
+        if(!found){
+            console.warn('Requested setting key to change: "'+ settingName+'" can not be found')
+        }else{
+            UiModeSystem.saveToLocalStorage()
+        }
+    }
+
+}
+
+export class UiMode {
+    private name : string;
+    private description : string;
+    private settingValues : SettingData[];
+    private isDefault : boolean;
+
+    constructor(name: string, description:string, settingValues: SettingData[], isDefault:boolean){
+        this.name = name;
+        this.description = description;
+        this.settingValues = settingValues;
+        this.isDefault=isDefault
+    }
+
+    getName = () : string => {
+        return this.name;
+    }
+
+    getDescription = () : string => {
+        return this.description;
+    }
+
+    getSettings = () : SettingData[] => {
+        return this.settingValues;
+    }
+
+    getIsDefault = () :boolean => {
+        return this.isDefault;
+    }
+
+    getSettingByKey = (key:string) : void => {
```

Comment:
Again maybe the for is very slightly better than forEach here and below in setSettingByKey, since the loops can terminate early when the setting is found.